docs/alternativesemantics.md

```diff
@@ -0,0 +1,387 @@
+This proposal can be adopted without resolving this design, but we should be careful not to restrict future optimisations and annotations.
+
+## Optimisations
+
+For this proposal to be successful, it must be possible to optimise code predictably. There are different constraints on optimising deallocation of regions to allowing early message sends.
+
+### Deallocation of regions
+
+Deallocation of regions is easier to optimise, as we can make the semantics less deterministic in the order that regions are deallocated. Consider the following code that creates to regions:
+```
+  var a = new C
+  var b = new C
+  a.f = new D in a
+  b.f = new D in b
+  var y = a.f;
+  drop a
+  drop b
+  // do something with y.
+```
+Now, the `drop a` will need to be delayed as `y` is keeping the region live, but `b` can be dropped instantly.  We can specify in the semantics that no particular order is guaranteed.  We could even extend systematic testing to delay non-deterministically deallocation any time it is valid to do so.
+
+#### Mechanisms for deciding.
+
+* Track potential internal pointers statically through abstract interpretation.
+  *  Keep set of pointers that may point into a particular region, and once that set is empty, can deallocate the region.
+  *  Taint regions, if we ever lose precision on if a region is still live, then mark region to be deallocated at the end of the scope/behaviour.
+* Track internal pointers dynamically.  We already have an additional root set that is used to enable tidy.  This set can be used to specify that there is an internal reference into the region associated to `a`, and once `y` goes away that could be removed.  This set is not expected to track references into sub-regions, but we could delay subregion deallocation further using the same mechanism.
+
+### Message Send
+
+Message send is much more complex as we are not allowed to reorder behaviours
+```
+when (a)
+{
+   // Captures something that is still live due to sub-reference
+}
+when (a)
+{
+  // Does not capture anything.
+}
+```
+It is tempting to immediately dispatch the second `when`, and only dispatch the first once it is safe to do so.  However, this breaks the causality guarantees of the language.
+
+Once we have delayed one `when`, then all subsequent `when`s must be delayed until we can dispatch the first.
+
+We could potentially attempt to track if we are break causality, but that requires us to be able to accurately determine may not alias on cowns:
+```
+when (a)
+{
+   // Captures something that is still live due to sub-reference
+}
+when (b)
+{
+  // Does not capture anything.
+}
+```
+It is sound to dispatch the second before the first, if `a` and `b` are different cowns, but otherwise it is not.
+
+The runtime if any message has been delayed, and if so, all subsequent messages in the behaviour are delayed until the end of the behaviour.  This is the simplist solution, and would probably work well in practice.

```

Comment:
Could you have a "depends on" pointer from one message to another if it "may depend on" and it's the compiler/runtime's job to optimise that away by checking alias etc?

Reply:
That would be an approach but would probably be too expensive to maintain.